File: util/MATLAB/loadTMP.m
% Load an ImageStack TMP file, an up to 4-dimensional array of integer or
% floating point data.
%
% output permutes the dimensions so that the output is:
% [height width channels frames]
%
% Format specification is here:
% https://github.com/mit-gfx/ImageStack/blob/master/src/FileTMP.cpp
function output = loadTMP(filename)

type_strings = {'*float32', '*float64', '*uint8', '*int8', ...
    '*uint16', '*int16', '*uint32', '*int32', '*uint64', '*int64'};

fid = fopen(filename, 'r');
header = fread(fid, 5, '*int32');

width = header(1);
height = header(2);
frames = header(3);
channels = header(4);
type_code = header(5);

if width < 1 || height < 1 || frames < 1 || channels < 1 || ...
        type_code < 0 || type_code > 9
    fclose(fid);
    error('Invalid header: %d %d %d %d %d', ...
        width, height, frames, channels, type_code);
end

num_elements = width * height * frames * channels;
type_string = type_strings{type_code + 1}; % MATLAB is one-based.

output = fread(fid, num_elements, type_string);
fclose(fid);

% Input is stored row major.
output = reshape(output, [width, height, frames, channels]);

% Permute dimensions so that each image is column major, and frames comes
% last.
output = permute(output, [2 1 4 3]);

File: util/numpy/loadTMP.py
# Load an ImageStack TMP file, an up to 4-dimensional array of integer or
# floating point data.
# Format specification is here:
# https://github.com/abadams/ImageStack/blob/master/src/FileTMP.cpp
#
# The output is permutted into the numpy C-contiguous convention: (frames, height, width, channels).
#   channels changes the fastest in memory, frames the slowest.
import numpy as np

def loadTMP(filename):
    with open(filename, 'rb') as f:
        buffer = f.read()

    type_code_list = \
    [
        np.dtype(np.float32),
        np.dtype(np.float64),
        np.dtype(np.uint8),
        np.dtype(np.int8),
        np.dtype(np.uint16),
        np.dtype(np.int16),
        np.dtype(np.uint32),
        np.dtype(np.int32),
        np.dtype(np.uint64),
        np.dtype(np.int64),
    ]

    offset = 0
    tmp_dims = np.frombuffer(buffer, count=4, dtype=np.int32)
    offset += tmp_dims.nbytes
    type_code = np.frombuffer(buffer, count=1, offset=offset, dtype=np.int32)
    offset += type_code.nbytes
    data_type = type_code_list[int(type_code)]

    # The TMP file writes the dimensions as:
    #   [width height frames channels]
    # Rewrite it in the numpy order:
    #   (channels, frames, height, width)
    a2_dims = np.flipud(tmp_dims)

    # Buffer stores the data as: (channels, frames, height, width).
    a2 = np.frombuffer(buffer, offset=offset, dtype=data_type)
    a2.shape = a2_dims

    # Permute the dimensions so that the data is returned in the numpy C-contiguous convention:
    #   (frames, height, width, channels).
    return a2.transpose((1, 2, 3, 0))

